server/auth/types/multiple/multi_auth.ts

```diff
@@ -0,0 +1,293 @@
+      wreckHttpsOption.checkServerIdentity = (host: string, cert: PeerCertificate) => {
+        return undefined;
+      };
+    }
+    if (Object.keys(wreckHttpsOption).length > 0) {
+      return wreck.defaults({
+        agents: {
+          http: new HTTP.Agent(),
+          https: new HTTPS.Agent(wreckHttpsOption),
+          httpsAllowUnauthorized: new HTTPS.Agent({
+            rejectUnauthorized: false,
+          }),
+        },
+      });
+    } else {
+      return wreck;
+    }
+  }
+
+  private async oidcConfigSetup(wreckClient: typeof wreck) {
+    const authHeaderName: string = this.config.openid?.header || '';
+    this.openIdAuthConfig.authHeaderName = authHeaderName;
+    const openIdConnectUrl: string = this.config.openid?.connect_url || '';
+
+    let scope = this.config.openid!.scope;
+    if (scope.indexOf('openid') < 0) {
+      scope = `openid ${scope}`;
+    }
+    this.openIdAuthConfig.scope = scope;
+
+    const response = await wreckClient.get(openIdConnectUrl);
+    const payload = JSON.parse(response.payload as string);
+
+    this.openIdAuthConfig.authorizationEndpoint = payload.authorization_endpoint;
+    this.openIdAuthConfig.tokenEndpoint = payload.token_endpoint;
+    this.openIdAuthConfig.endSessionEndpoint = payload.end_session_endpoint;
+  }
+
+  async isValidCookie(cookie: SecuritySessionCookie): Promise<boolean> {
+    if (cookie.authType === AuthType.BASIC) {
+      const result =
+        cookie.expiryTime &&
+        ((cookie.username && cookie.credentials?.authHeaderValue) ||
+          (this.config.auth.anonymous_auth_enabled && cookie.isAnonymousAuth));
+
+      return result;
+    } else if (cookie.authType === AuthType.OPEN_ID) {
+      if (
+        !cookie.username ||
+        !cookie.expiryTime ||
+        !cookie.credentials?.authHeaderValue ||
+        !cookie.credentials?.expires_at
+      ) {
+        return false;
+      }
+      if (cookie.credentials?.expires_at > Date.now()) {
+        return true;
+      }
+
+      // need to renew id token
+      if (cookie.credentials.refresh_token) {
+        try {
+          const query: any = {
+            grant_type: 'refresh_token',
+            client_id: this.config.openid?.client_id,
+            client_secret: this.config.openid?.client_secret,
+            refresh_token: cookie.credentials.refresh_token,
+          };
+          const refreshTokenResponse = await callTokenEndpoint(
+            this.openIdAuthConfig.tokenEndpoint!,
+            query,
+            this.wreckClient
+          );
+
+          // if no id_token from refresh token call, maybe the Idp doesn't allow refresh id_token
+          if (refreshTokenResponse.idToken) {
+            cookie.credentials = {
+              authHeaderValue: `Bearer ${refreshTokenResponse.idToken}`,
+              refresh_token: refreshTokenResponse.refreshToken,
+              expires_at: Date.now() + refreshTokenResponse.expiresIn! * 1000, // expiresIn is in second
```

Comment:
Why is credential valid for only 1 second?

Reply:
how did you come up with this 1 second?

Reply:
My bad. I misread the comment here.